spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/MessageSerializationUtils.java

```diff
@@ -0,0 +1,171 @@
+import java.io.UnsupportedEncodingException;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.core.serializer.support.SerializationFailedException;
+import org.springframework.integration.codec.Codec;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.ContentTypeResolver;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.MimeType;
+import org.springframework.util.MimeTypeUtils;
+import org.springframework.util.ObjectUtils;
+
+/**
+ * Utility class for serializing and de-serializing the message payload.
+ *
+ * @author Soby Chacko
+ */
+public abstract class MessageSerializationUtils {
+
+	protected static final Log LOGGER = LogFactory.getLog(MessageSerializationUtils.class);
+
+	/**
+	 * Serialize the message payload unless it is a byte array.
+	 *
+	 * @param message the message with the payload to serialize
+	 * @param codec the codec used for serialization
+	 * @return the Message with teh serialized payload
+	 */
+	public static MessageValues serializePayload(Message<?> message, Codec codec) {
+		Object originalPayload = message.getPayload();
+		Object originalContentType = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
+
+		// Pass content type as String since some transport adapters will exclude
+		// CONTENT_TYPE Header otherwise
+		Object contentType = JavaClassMimeTypeUtils
+				.mimeTypeFromObject(originalPayload, ObjectUtils.nullSafeToString(originalContentType)).toString();
+		Object payload = serializePayload(originalPayload, codec);
+		MessageValues messageValues = new MessageValues(message);
+		messageValues.setPayload(payload);
+		messageValues.put(MessageHeaders.CONTENT_TYPE, contentType);
+		if (originalContentType != null && !originalContentType.toString().equals(contentType.toString())) {
+			messageValues.put(BinderHeaders.BINDER_ORIGINAL_CONTENT_TYPE, originalContentType.toString());
+		}
+		return messageValues;
+	}
+
+	/**
+	 * Serialize the payload object if it is not a byte array.
+	 *
+	 * @param originalPayload the payload to serialize
+	 * @param codec the codec used for serialization
+	 * @return the serialized byte array or the original payload if it is already a byte array
+	 */
+	public static byte[] serializePayload(Object originalPayload, Codec codec) {
+		if (originalPayload instanceof byte[]) {
+			return (byte[]) originalPayload;
+		}
+		else {
+			ByteArrayOutputStream bos = new ByteArrayOutputStream();
+			try {
+				if (originalPayload instanceof String) {
+					return ((String) originalPayload).getBytes("UTF-8");
+				}
+				codec.encode(originalPayload, bos);
+				return bos.toByteArray();
+			}
+			catch (IOException e) {
+				throw new SerializationFailedException(
+						"unable to serialize payload [" + originalPayload.getClass().getName() + "]", e);
```

Comment:
Would be better to mention in the JavaDoc what is the exception this method is going to throw and why